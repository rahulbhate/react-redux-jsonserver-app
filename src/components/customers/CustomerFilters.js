import React from "react";
import { connect } from "react-redux";
import DropDown from "../common/Dropdown";
import Label from "../common/Label";
import TextInput from "../common/TextInput";
import { filterText, sortBy } from "../../redux/actions/filters";
const arrays = ["firstname", "lastname"];
const CustomerFilters = props => (
  <div className='measure center'>
    <TextInput
      htmlId='example-optional'
      label='Search By Customer First Name and Last Name'
      placeholder='Search...'
      name='search'
      onChange={e => {
        props.dispatch(filterText(e.target.value));
      }}
    />
    <Label htmlFor='test' label='SORT BY' />
    <DropDown
      id='mySelection'
      name='customers'
      htmlFor='customers'
      placeholder='SORT BY'
      options={arrays}
      handleChange={e => props.dispatch(sortBy(e.target.value))}
      className='form-control'
    ></DropDown>
  </div>
);

const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};

export default connect(mapStateToProps)(CustomerFilters);
