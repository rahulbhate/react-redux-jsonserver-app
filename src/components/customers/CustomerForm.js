import React from "react";
import PropTypes from "prop-types";
import TextInput from "../common/TextInput";
import Button from '../common/Button';
import Label from '../common/Label';
const CustomerForm = ({
  customer,
  onSave,
  onChange,
  saving = false,
  errors = {}
}) => {
  return (
    <form onSubmit={onSave} className='measure center'>
      <h2>{customer.id ? "Edit" : "Add"} Customer</h2>
      {errors.onSave && (
        <div className='alert alert-danger' role='alert'>
          {errors.onSave}
        </div>
      )}
      <Label htmlFor="firstname" label="First Name" required />
      <TextInput
        name='firstName'
        value={customer.firstName}
        onChange={onChange}
        error={errors.firstname}
      />
      <Label htmlFor="lastname" label="Last Name" required />
      <TextInput
        name='lastName'
        value={customer.lastName}
        onChange={onChange}
        error={errors.lastname}
      />
      <Label htmlFor="birthdate" label="Birth Date" required />
      <TextInput
        name='birthDate'
        type="date"
        value={customer.birthDate}
        onChange={onChange}
        error={errors.birthdate}
      />
      <Button type='submit' disabled={saving} className='btn btn-primary' title={saving ? "Saving..." : "Save"} />
    </form>
  );
};

CustomerForm.propTypes = {
  customer: PropTypes.object.isRequired,
  errors: PropTypes.object,
  onSave: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  saving: PropTypes.bool
};

export default CustomerForm;
