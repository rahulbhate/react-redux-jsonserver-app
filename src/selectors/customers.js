const getCustomers = (customers, { text, sortBy }) => {
  return  customers.length === 0
    ? []
    : customers
        .map(customer => {
          return {
            ...customer,
          };
        })
        .filter(customer => {
          const textMatch =
            customer.firstName.toLowerCase().includes(text.toLowerCase()) ||
            customer.lastName.toLowerCase().includes(text.toLowerCase());
          return textMatch;
        })
        .sort((cust1, cust2) => {
          if (sortBy === "firstname") {
            return cust1.firstName.localeCompare(cust2.firstName);
          } else if (sortBy === "lastname") {
            return cust1.lastName.localeCompare(cust2.lastName);
          }
        });
};
export default getCustomers;
