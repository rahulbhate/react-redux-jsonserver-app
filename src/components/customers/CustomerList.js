import React from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import CustomerFilters from "./CustomerFilters";
const CustomerList = ({ customers, onDeleteClick }) => {
  return (
    <>
      <CustomerFilters />
      <table className='table table-bordered table-light table-striped'>
        <thead>
          <tr>
            <th>First Name</th>
            <th>Last Name</th>
            <th>Birth Date</th>
            <th />
          </tr>
        </thead>
        <tbody>
          {customers.map(customer => {
            return (
              <tr key={customer.id}>
                <td>
                  <Link to={"/customer/" + customer.firstName}>{customer.firstName}</Link>
                </td>
                <td>
                <Link to={"/customer/" + customer.firstName}>{customer.lastName}</Link>
                </td>
                <td>{customer.birthDate}</td>
                <td>
                  <button
                    className='btn btn-outline-danger'
                    onClick={() => onDeleteClick(customer)}
                  >
                    Delete
                  </button>
                </td>
              </tr>
            );
          })}
        </tbody>
      </table>
    </>
  );
};

CustomerList.propTypes = {
  customers: PropTypes.array.isRequired,
  onDeleteClick: PropTypes.func.isRequired
};

export default CustomerList;
