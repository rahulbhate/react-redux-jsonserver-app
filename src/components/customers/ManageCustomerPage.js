import React, { useState, useEffect } from "react";
import { connect } from "react-redux";
import { loadCustomers, saveCustomer } from "../../redux/actions/customerActions";
import PropTypes from "prop-types";
import CustomerForm from "./CustomerForm";
import Spinner from "../common/Spinner";
import { toast } from "react-toastify";
import { newCustomer } from "../../../tools/mockData";
const ManageCustomerPage = ({
  customers,
  loadCustomers,
  saveCustomer,
  history,
  ...props
}) => {
  const [customer, setCustomer] = useState({ ...props.customer });
  const [errors, setErrors] = useState({});
  const [saving, setSaving] = useState(false);

  useEffect(() => {
    console.log("Use Effect Hook Called");
    if (customers.length === 0) {
      loadCustomers().catch(error => {
        alert("Loading customers failed" + error);
      });
    } else {
      setCustomer({ ...props.customer });
    }

  }, [props.customer]);
  function handleChange(event) {
    const { name, value } = event.target;
    setCustomer(prevCustomer => ({
      ...prevCustomer,
      [name]:
        name === "authorId" || name === "categoryId"
          ? parseInt(value, 10)
          : value
    }));
    console.log(customer);
  }
  
  function formIsValid() {
    const { firstName, lastName, birthDate } = customer;
    const errors = {};

    if (!firstName) errors.firstname = "First Name is required.";
    if (!lastName) errors.lastname = "Last Name is required";
    if (!birthDate) errors.birthdate = "Birth Date is required";
    setErrors(errors);
    // Form is valid if the errors object still has no properties
    return Object.keys(errors).length === 0;
  }

  function handleSave() {
    event.preventDefault();
    if (!formIsValid()) return;
      setSaving(true);
      saveCustomer(customer)
        .then(() => {
          toast.info("Customers Saved Successfully");
          history.push("/customers");
        })
        .catch(error => {
          setSaving(false);
          setErrors({ onSave: error.message });
        });
  }
  return customers.length === 0  ? (
    <Spinner />
  ) : (
    <div className='mw6 center bg-white br3 pa3 pa5-ns mv5 ba b--black-7'>
      <CustomerForm
        customer={customer}
        errors={errors}
        onChange={handleChange}
        onSave={handleSave}
        saving={saving}
      />
    </div>
  );
};
export function getCustomersBySlug(customer, slug) {
  return customer.find(customer => customer.firstName === slug) || null;
}
const mapStateToProps = (state, ownProps) => {
  const slug = ownProps.match.params.slug;
  console.log(slug);
  const customer =
    slug && state.customers.length > 0
      ? getCustomersBySlug(state.customers, slug)
      : newCustomer;
  return {
    customer,
    customers: state.customers,
  };
};
const mapDispatchToProps = {
  loadCustomers,
  saveCustomer,
};
ManageCustomerPage.propTypes = {
  customers: PropTypes.array.isRequired,
  loadCustomers: PropTypes.func.isRequired,
  saveCustomer: PropTypes.func.isRequired,
  history: PropTypes.object.isRequired
};
export default connect(mapStateToProps, mapDispatchToProps)(ManageCustomerPage);
