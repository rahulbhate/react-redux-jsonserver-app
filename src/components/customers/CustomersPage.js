import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import {
  loadCustomers,
  deleteCustomer,
  searchCustomer
} from "../../redux/actions/customerActions";
import getVisibleCustomers from "../../selectors/customers";
import PropTypes from "prop-types";
import CustomerList from "./CustomerList";
import Spinner from "../common/Spinner";
import Button from "../common/Button";
import { toast } from "react-toastify";
import { Redirect } from "react-router-dom";
import { getCustomers } from "../../api/customerApi";
const CustomersPage = ({
  loadCustomers,
  customers,
  deleteCustomer,
  searchCustomer,
  ...props
}) => {
  const [redirectToAddCustomerPage, setRedirectToAddCustomerPage] = useState(false);
  const [loading, setLoading] = useState(false);
  useEffect(() => {
    console.log("Use Effect Hook Called");
    if (customers.length === 0) {
      loadCustomers().catch(error => {
        alert("Loading customers failed" + error);
      });
    }
  }, []);

  const handleDeleteCourse = async customer => {
    toast.info("Customer deleted");
    try {
      await deleteCustomer(customer);
    } catch (error) {
      toast.error("Delete failed. " + error.message, { autoClose: false });
    }
  };
  return (
    <>
      {redirectToAddCustomerPage && <Redirect to='/customer' />}
      {props.loading ? (
        <Spinner />
      ) : (
        <>
          <div className='fl w-100 bg-near-white tc'>
            <Button
              title='Add Course'
              className='f6 no-underline grow dib v-mid bg-purple white ba b--light-purple ph3 pv2 mb3'
              type='primary'
              onClick={() => {
                setRedirectToAddCustomerPage({ redirectToAddCustomerPage: true });
              }}
            />
            <CustomerList onDeleteClick={handleDeleteCourse} customers={customers} />
          </div>
        </>
      )}
    </>
  );
};

const mapStateToProps = state => {
  return {
    customers: getVisibleCustomers(
      state.customers,
      state.filters
    ),
    //courses: selectedCoursesSelector(state),
    loading: state.apiCallsInProgress > 0
  };
};
const mapDispatchToProps = {
  loadCustomers,
  deleteCustomer,
  searchCustomer
};
CustomersPage.propTypes = {
  customers: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  loadCustomers: PropTypes.func.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(CustomersPage);
