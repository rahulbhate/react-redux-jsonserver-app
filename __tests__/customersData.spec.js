import customers from '../tools/db.json';
test('customers data is correct', () => {
    expect(customers).toMatchSnapshot();
    expect(1).toEqual(1);
});
// test each item in the customers data has the correct properties
for(let i=0; i<customers.length; i++){
    test(`customer[$(i)] should have properties (id,firstName,lastName,birthDate,createdAt)`, () =>{
        expect(customer[i]).toHaveProperty('id');
        expect(customer[i]).toHaveProperty('firstName');
        expect(customer[i]).toHaveProperty('lastName');
        expect(customer[i]).toHaveProperty('birthDate');
        expect(customer[i]).toHaveProperty('createdAt');
    });
}
// default jest mock function
test('mock implementation of a basic function', () =>{
    const mock = jest.fn(() => 'I am a mock function');
    console.log(mock);
    expect(mock('Calling my mock function!')).toBe('I am a mock function');
    expect(mock).toHaveBeenCalledWith('Calling my mock function!');
});

// let's mock the return value and test calls
test('mock return value of a function one time', () => {
    const mock = jest.fn();
  
    // we can chain these!
    mock.mockReturnValueOnce('Hello').mockReturnValueOnce('there!');
  
    mock(); // first call 'Hello'
    mock(); // second call 'there!'
  
    expect(mock).toHaveBeenCalledTimes(2); // we know it's been called two times
  
    mock('Hello', 'there', 'Steve'); // call it with 3 different arguments
    expect(mock).toHaveBeenCalledWith('Hello', 'there', 'Steve');
  
    mock('Steve'); // called with 1 argument
    expect(mock).toHaveBeenLastCalledWith('Steve');
  });

  // let's mock the return value
// difference between mockReturnValue & mockImplementation
test('mock implementation of a function', () => {
    const mock = jest.fn().mockImplementation(() => 'Australia');
    expect(mock('Location')).toBe('Australia');
    expect(mock).toHaveBeenCalledWith('Location');
  });


// async example, always return a promise (can switch out resolves with reject)
test('expect a promise to resolve', async () => {
    const user = {
      getFullName: jest.fn(() => Promise.resolve('Rahul Bhate')),
    };
    await expect(user.getFullName('Rahul Bhate')).resolves.toBe('Rahul Bhate');
  });
  
  test('expect a promise to reject', async () => {
    const user = {
      getFullName: jest.fn(() =>
        Promise.reject(new Error('Something went wrong'))
      ),
    };
    await expect(user.getFullName('Rahul Bhate')).rejects.toThrow(
      'Something went wrong'
    );
  });