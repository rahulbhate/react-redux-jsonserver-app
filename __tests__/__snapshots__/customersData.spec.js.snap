// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customers data is correct 1`] = `
Object {
  "customers": Array [
    Object {
      "birthDate": "2020-10-08",
      "createdAt": 1598778507721,
      "firstName": "Cory",
      "id": 10,
      "lastName": "House",
    },
    Object {
      "birthDate": "2020-10-09",
      "createdAt": 1602212313220,
      "firstName": "Martin",
      "id": 11,
      "lastName": "Shub",
    },
    Object {
      "birthDate": "2010-06-09",
      "createdAt": 1602244811500,
      "firstName": "John",
      "id": 12,
      "lastName": "Crow",
    },
    Object {
      "birthDate": "2006-02-22",
      "createdAt": 1602244966371,
      "firstName": "Jack",
      "id": 13,
      "lastName": "Ramen",
    },
  ],
}
`;
